File: frontend/src/types/contracts/v2/asyncLoop.ts
/** Async creation loop contracts (Spec-013 Phase 1). */

export type InboxKind = 'text' | 'image' | 'voice' | 'link' | 'idea';
export type InboxConsent = 'publishable' | 'private';
export type ContentIntent = 'solve' | 'share' | 'record';
export type DeliverableStatus =
  | 'queued'
  | 'producing'
  | 'ready'
  | 'failed'
  | 'expired'
  | 'picked'
  | 'discarded';

export interface InboxAddInput {
  kind: InboxKind;
  title?: string;
  content: string;
  consent?: InboxConsent;
  idempotency_key: string;
}

export interface InboxItem {
  id: string;
  kind: InboxKind;
  title: string;
  content: string;
  consent: InboxConsent;
  status: 'intake' | 'digested' | 'failed';
  version: number;
  created_at: string;
  updated_at: string;
}

export interface DeliverableFact {
  statement: string;
  source_inbox_id: string;
  note?: string;
}

export interface DeliverableJudgment {
  audience_change?: string;
  primary_response?: string;
  supporting?: string[];
  window_days?: number;
}

export interface Deliverable {
  id: string;
  thread_id: string;
  title: string;
  body_text: string;
  outline: { step: string; label: string }[];
  facts: DeliverableFact[];
  judgment: DeliverableJudgment;
  content_intent: ContentIntent | null;
  proposed_publish_at: string | null;
  is_exploration: boolean;
  status: DeliverableStatus;
  attribution: string | null;
  expire_at: string | null;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface PickupInput {
  content_intent: ContentIntent;
  audience_change: string;
  primary_response?: 'save' | 'comment' | 'profile_visit' | 'follow';
  window_days?: number;
  schedule_at?: string;
  idempotency_key: string;
}

export interface DiscardInput {
  reason: string;
  idempotency_key: string;
}

export interface PickupResult {
  project: { id: string; title: string };
  deliverable: Deliverable;
}

export interface DigestResult {
  thread_id: string;
  deliverables: Deliverable[];
}

export interface MetricRecord {
  id: string;
  metric: 'pickup_seconds' | 'weekly_minutes' | 'published_count' | 'discard_attribution';
  value: number;
  meta: Record<string, unknown>;
  created_at: string;
}

File: backend/app/models/v2/async_loop.py
"""Typed contracts for the async creation loop (Spec-013 Phase 1)."""

from typing import Any, Literal

from pydantic import Field

from app.models.v2.intent_actions import StrictModel

InboxKind = Literal["text", "image", "voice", "link", "idea"]
InboxConsent = Literal["publishable", "private"]


class InboxItemCreate(StrictModel):
    kind: InboxKind
    title: str = Field(default="", max_length=120)
    content: str = Field(min_length=1, max_length=4000)
    consent: InboxConsent = "publishable"
    idempotency_key: str = Field(min_length=1, max_length=200)


class InboxItemView(StrictModel):
    id: str
    kind: InboxKind
    title: str
    content: str
    consent: InboxConsent
    status: Literal["intake", "digested", "failed"]
    version: int
    created_at: str
    updated_at: str


class DeliverableView(StrictModel):
    id: str
    thread_id: str
    title: str
    body_text: str
    outline: list[Any]
    facts: list[Any]
    judgment: dict[str, Any]
    content_intent: Literal["solve", "share", "record"] | None
    proposed_publish_at: str | None
    is_exploration: bool
    status: Literal[
        "queued", "producing", "ready", "failed", "expired", "picked", "discarded"
    ]
    attribution: str | None
    expire_at: str | None
    precheck: dict[str, Any] = Field(default_factory=dict)
    version: int
    created_at: str
    updated_at: str


class PickupRequest(StrictModel):
    content_intent: Literal["solve", "share", "record"]
    audience_change: str = Field(min_length=1, max_length=1000)
    primary_response: Literal["save", "comment", "profile_visit", "follow"] = "save"
    window_days: int = Field(default=7, ge=1, le=365)
    schedule_at: str | None = Field(default=None, max_length=40)
    idempotency_key: str = Field(min_length=1, max_length=200)


class DiscardRequest(StrictModel):
    reason: str = Field(min_length=1, max_length=60)
    idempotency_key: str = Field(min_length=1, max_length=200)


class MetricsRecord(StrictModel):
    metric: Literal[
        "pickup_seconds", "weekly_minutes", "published_count", "discard_attribution"
    ]
    value: float
    meta: dict[str, Any] = Field(default_factory=dict)
